plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection.php

```diff
@@ -1980,11 +1980,26 @@ function ( $collection_args ) {
 					);
 				}
 
+				$category_callback = function () use ( $collection_args ) {
+					return $collection_args['relatedBy']['categories'];
+				};
+
+				$tag_callback = function () use ( $collection_args ) {
+					return $collection_args['relatedBy']['tags'];
+				};
+
+				add_filter( 'woocommerce_product_related_posts_relate_by_category', $category_callback, PHP_INT_MAX );
+				add_filter( 'woocommerce_product_related_posts_relate_by_tag', $tag_callback, PHP_INT_MAX );
+
 				$related_products = wc_get_related_products(
 					$collection_args['relatedProductReference'],
 					// Use a higher limit so that the result set contains enough products for the collection to subsequently filter.
 					100
 				);
+
+				remove_filter( 'woocommerce_product_related_posts_relate_by_category', $category_callback, PHP_INT_MAX );
+				remove_filter( 'woocommerce_product_related_posts_relate_by_tag', $tag_callback, PHP_INT_MAX );
```

Comment:
I'm using WooCommerce's filter hooks to control which taxonomies (categories/tags) are used for finding related products, then removing the filters after getting results. I chose this over creating custom queries to leverage WooCommerce's existing optimized logic.

Does this approach make sense? Any concerns about using filters this way?